test/dist_autograd_test.py

```diff
@@ -60,63 +78,141 @@ def test_autograd_context(self):
+        )
+        self.assertEqual(
+            "torch::distributed::autograd::RecvRpcBackward", next_funcs[1][0].name()
+        )
+        self.assertEqual(next_funcs[0][0], next_funcs[1][0])
+
+    def _test_autograd_functions(self, fn):
         dst_rank = (self.rank + 1) % self.world_size
         with dist_autograd.context() as context_id:
             t1 = torch.ones(3, 3, requires_grad=True)
             t2 = torch.zeros(3, 3, requires_grad=True)
-            ret = rpc.rpc_sync("worker{}".format(dst_rank), torch.add, args=(t1, t2))
-            rpc.rpc_sync(
-                "worker{}".format(dst_rank), _set_rpc_done, args=(context_id,)
-            )
+            ret = rpc.rpc_sync("worker{}".format(dst_rank), fn, args=(t1, t2))
+            rpc.rpc_sync("worker{}".format(dst_rank),
+                         _set_rpc_done_from_prev_rank, args=(context_id,))
 
-            # Get send function.
+            self._verify_send_recv_functions_in_client(context_id, t1, t2, ret)
+
+            # We should have send/recv functions from the previous rank, get all
+            # contexts in this node to find them.
+
+            # Wait for the prev rank to be done with rpc.
+            while not prev_rank_rpc_done:
+                time.sleep(0.1)
+                pass
+
+            # Now verify the autograd graph.
+            ctx = dist_autograd._retrieve_context(prev_rank_context_id)
+
+            self._verify_send_recv_functions_in_tensor_run(ctx)
+
+        # autograd context should be cleaned up by now.
+        with self.assertRaises(RuntimeError):
+            ctx = dist_autograd._retrieve_context(context_id)
+
+        # No autograd context available.
+        with self.assertRaises(RuntimeError):
             ctx = dist_autograd._current_context()
-            self.assertEqual(context_id, ctx._context_id())
-            send_functions = ctx._send_functions()
-            self.assertEqual(1, len(send_functions))
 
-            # Retrieve the next functions in the graph.
-            next_funcs = list(send_functions.values())[0].next_functions
-            self.assertEqual(2, len(next_funcs))
+    @dist_init
+    def test_autograd_functions_for_builtin_call(self):
+        self._test_autograd_functions(torch.add)
+
+    @dist_init
+    def test_autograd_functions_for_python_call(self):
+        self._test_autograd_functions(my_py_add)
 
-            # We should now hit t1 and t2 in the autograd graph.
-            self.assertEqual("torch::autograd::AccumulateGrad", next_funcs[0][0].name())
-            self.assertEqual(t1, next_funcs[0][0].variable)
-            self.assertEqual(0, next_funcs[0][1])
-            self.assertEqual("torch::autograd::AccumulateGrad", next_funcs[1][0].name())
-            self.assertEqual(t2, next_funcs[1][0].variable)
-            self.assertEqual(0, next_funcs[1][1])
+    @dist_init
+    def test_autograd_functions_for_python_nested_call(self):
```

Comment:
Can we add a test where the nested RPC calls itself? Basically A->B->A?

Reply:
Also, can we add a test for more than 1 layer of nested calls? (could extend this test itself too). Something like A->B->C->D.

Reply:
will add them